### ISSUE TITLE
Issue occurred when loop export GraphicLayoutWidget without showing it

### ISSUE THREAD

**Issue by shixi2ng** [pyqtgraph/pyqtgraph#1997]
Sorry for disturbing the contributors, I am not here to report a bug but find for a help, as at this moment I cannot find any other ways to solve my problem. The primary aim of my program is to generate about 25,000 figures with 20 subplots in each of them. As I found out matplotlib is not the best choice to loop export so many figures, I turn to use pyqtgraph. But I encountered two issues: 
(p1) Without showing the figure, the code exports the GraphicLayoutWidget with all subplots stacked in the upper-left corner of the window as shown in the figure below.
<img width="1250" alt="annual_NDVI_259_57" src="https://user-images.githubusercontent.com/51545550/134182872-8c295f60-d367-4501-af6e-1df2ee5011e2.png">
(p2)I try to fix (p1) by manually setting the range of the graphic view, the exported png looks fine whereas the scatterplot seems to have an offset. 
![741632232592_ pic_hd](https://user-images.githubusercontent.com/51545550/134185005-caa80f44-9204-46b6-bccc-986784198a43.jpg)
(As you can see the scatterplot has an offset compared to the figure below)
![751632232627_ pic_hd](https://user-images.githubusercontent.com/51545550/134185027-b220e3bd-9d7a-4651-aa83-7a6e3bd24452.jpg)
(This is the correct one after I show it)
Someone mentioned this issue before, after I explored all these discussions, I set the PXmode as True in the scatterplot. It truly works but since the local x y coordinate is not on the same scale so all the points were either like a needle or only a pixel small. 
![731632232544_ pic](https://user-images.githubusercontent.com/51545550/134185654-d3a243a8-741f-47b8-b134-2555e8a49c72.jpg)
So can someone help me with these issues, or just give me some useful links or examples?
### Code to reproduce

```python
import pyqtgraph as pg
import numpy as np


x_temp = np.linespace(0, 365, 10000)
y_temp = np.linespace(0, 1, 10000)
rows = 5
columns = 5
year_range = range(2000, 2021)
pg.setConfigOption('background', 'w')
# pg.setConfigOptions(antialias=True)
win = pg.GraphicsLayoutWidget(show=False, title="Annual phenology")
win.setRange(newRect=pg.Qt.QtCore.QRectF(140, 100, 500 * columns-200, 300 * rows-200), disableAutoPixel=False)
win.resize(500 * columns, 300 * rows)
year_t = 0
line_pen = pg.mkPen((0, 0, 255), width=5)
# x_tick_num = [15, 44, 75, 105, 136, 166, 197, 228, 258, 289, 320, 351]
# x_tick_label = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
# x_tick = [[x_tick_num[i], x_tick_label[i]] for i in range(len(x_tick_num))]
x_tick = [list(zip((15, 44, 75, 105, 136, 166, 197, 228, 258, 289, 320, 351), ('Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec')))]
for r_temp in range(rows):
    for c_temp in range(columns):
        if year_t < len(year_range):
            year = year_range[year_t]   
            R_square = 50.000
            msg_r_square = (r'R^2 = ' + str(R_square)[0:5] + '%')
            plot_temp = win.addPlot(row=r_temp, col=c_temp, title='Annual phenology of Year ' + str(year))
            plot_temp.setLabel('left', 'phenology index')
            plot_temp.setLabel('bottom', 'DOY')
            scatter_array = np.stack((x_temp, y_temp), axis=1)
            s1 = pg.ScatterPlotItem(size=5, pxMode=True)
            s1.setData(scatter_array[:,0], scatter_array[:,1])
            s1.setPen(pg.mkPen('r', width=10))
            s1.setBrush(pg.mkBrush(255, 0, 0))
            plot_temp.addItem(s1)
            plot_temp.plot(x_temp, y_temp, pen=line_pen, name="Phenology_index")
            t1 = pg.TextItem(text=msg_r_square)
            t1.setPos(260, 0.92)
            plot_temp.addItem(t1)
            x_axis = plot_temp.getAxis('bottom')
            x_axis.setTicks(x_tick)
        year_t += 1
# win.show()
exporter = pg.exporters.ImageExporter(win.scene())
# exporter.parameters()['width'] = 500 * columns
exporter.export(phenology_fig_dic['individual_curve_path'] + 'annual_' + str(vi) + '_' + str(x) + '_' + str(y) + '.png')


### Tested environment(s)

 * PyQtGraph version:  0.11.0
 * Qt Python binding: PyQt5 5.9.2 Qt 5.9.6
 * Python version:  3.6
 * NumPy version: 1.17.0
 * Operating system: win10
 * Installation method: conda

**Comment by NilsNemitz** [comment]
Hi @shixi2ng , that's an impressively scaled project you have there!

I am not really an expert on how the overall layouting process works, but there are indeed multiple steps of communication between the various Qt and pyqtgraph elements that need to happen before the layout is complete.

If you need to make sure that the layout happens as quickly as possible, you should probably try to make it as simple as possible, and work with some widget of fixed size, so that -at least- there is no negotiation about how to arrange multiple items of initially unknown size.

Some questions and suggestions:

Do all the 25,000 figures share the same layout?
Do you need to set up the layout for each of the figures?
I would expect it to be much faster to keep the layout the same. Keep a list of the 20 ScatterPlotItems. Update by calling `setData` on all of them for each of the 25,000 plots. That may be what you are doing already, but I could not identify that from your example code, sorry.

Is displaying the plots still slow when you do that?
What amount of time does your current output take with and without display? 
What time do you need to achieve?
Usually, pyqtgraph can manage update rates of >100/s for such plots with a reasonable amount of data, so the "display new data" step should not take more than a few minutes, even for 25,000 plots.

Other things to try:
Disable auto range and manually set the axes ranges. Auto-range causes an additional redraw under some circumstances, which might be slowing things down if you are going for *really* high update rates.
And maybe try setting all the formatting options before you add the data. By design, nothing should get drawn until Qt requests an update to the display. But ScatterPlotItem has a lot of caching, and it is *possible* that setting the data first and then formatting may invalidate and regenerate some cached data. If you find that this makes a difference, please let us know!

**Comment by shixi2ng** [comment]
Hi @NilsNemitz, Thanks for your comprehensive reply!
Since I got into this trouble, I set up some basic timers within my code to check which part was consuming the most. Consequently, I found out the render time for each individual image was sequentially slow down and apparently, something went wrong between each loop. Based on this information, I got some useful links, like [](https://stackoverflow.com/questions/35535830/how-to-clear-a-plot-in-a-while-loop-when-using-pyqtgraph), which all refer to that loop creating the PlotItem consumes brunch of time and I finally got a clear picture of how the pyqtgraph works on plotting figures. Thus, I change my code as is shown down below, I will fully paste here hope it can help someone in the future. The result is outstanding, for each image, the rendering time decreases from 3s to 0.5s approximately where .4s is used to executing the curve fitting section 20 times. So, I think everything is fine for me right now! Again, I truly appreciate your reply and respect to all your contributors here. If you don't have anything else to add, I think we can close this issue.

    if initial_t:
        phe_dic = {}
        win = pg.GraphicsLayoutWidget(show=False, title="Annual phenology")
        win.setRange(newRect=pg.Qt.QtCore.QRectF(140, 100, 500 * columns-200, 300 * rows-200), disableAutoPixel=False)
        win.resize(500 * columns, 300 * rows)
        year_t = 0
        for r_temp in range(rows):
            for c_temp in range(columns):
                if year_t < len(year_range):
                    year = year_range[year_t]
                    phe_dic['plot_temp_' + str(year)] = win.addPlot(row=r_temp, col=c_temp, title='Annual phenology of Year ' + str(year))
                    phe_dic['plot_temp_' + str(year)].setLabel('left', vi)
                    phe_dic['plot_temp_' + str(year)].setLabel('bottom', 'DOY')
                    x_axis = phe_dic['plot_temp_' + str(year)].getAxis('bottom')
                    x_axis.setTicks(x_tick)
                    phe_dic['curve_temp_' + str(year)] = pg.PlotCurveItem(pen=line_pen, name="Phenology_index")
                    phe_dic['plot_temp_' + str(year)].addItem(phe_dic['curve_temp_' + str(year)])
                    phe_dic['plot_temp_' + str(year)].setRange(xRange=(0, 365), yRange=(0, 0.95))
                    phe_dic['scatterplot_temp_' + str(year)] = pg.ScatterPlotItem(size=0.01, pxMode=False)
                    phe_dic['scatterplot_temp_' + str(year)].setPen(pg.mkPen('r', width=10))
                    phe_dic['scatterplot_temp_' + str(year)].setBrush(pg.mkBrush(255, 0, 0))
                    phe_dic['plot_temp_' + str(year)].addItem(phe_dic['scatterplot_temp_' + str(year)])
                    phe_dic['text_temp_' + str(year)] = pg.TextItem()
                    phe_dic['text_temp_' + str(year)].setPos(260, 0.92)
                    phe_dic['plot_temp_' + str(year)].addItem(phe_dic['text_temp_' + str(year)])
                year_t += 1
        initial_t = False

        year_t = 0
        for r_temp in range(rows):
            for c_temp in range(columns):
                if year_t < len(year_range):
                    year = year_range[year_t]
                    if np.argwhere(year_temp_temp == year).shape[0] == 0:
                        phe_dic['curve_temp_' + str(year)].setData(np.array([-2, -1]), np.array([-2, -1]))
                        phe_dic['text_temp_' + str(year)].setText('')
                        phe_dic['scatterplot_temp_' + str(year)].setData(np.array([-2, -1]), np.array([-2, -1]))
                    else:
                        phase1_s = time.time()
                        p_min = np.min(np.argwhere(year_temp_temp == year))
                        p_max = np.max(np.argwhere(year_temp_temp == year)) + 1
                        annual_doy_temp = doy_temp_temp[p_min: p_max]
                        annual_phenology_index_temp = phenology_index_temp[p_min: p_max]
                        # plot_temp.enableAutoRange()
                        phase1_time += time.time() - phase1_s
                        phase2_s = time.time()
                        scatter_array = np.stack((annual_doy_temp, annual_phenology_index_temp), axis=1)
                        phe_dic['scatterplot_temp_' + str(year)].setData(scatter_array[:, 0], scatter_array[:, 1])
                        phase2_time += time.time() - phase2_s
                        phase3_s = time.time()
                        if annual_phenology_index_temp.shape[0] >= 7:
                            paras, extras = curve_fit(curve_fitting_algorithm, annual_doy_temp, annual_phenology_index_temp, maxfev=5000, p0=VI_curve_fitting['para_ori'], bounds=VI_curve_fitting['para_boundary'])
                            predicted_phenology_index = seven_para_logistic_function(annual_doy_temp, paras[0], paras[1], paras[2], paras[3], paras[4], paras[5], paras[6])
                            R_square = (1 - np.sum((predicted_phenology_index - annual_phenology_index_temp) ** 2) / np.sum((annual_phenology_index_temp - np.mean(annual_phenology_index_temp)) ** 2)) * 100
                            msg_r_square = (r'R^2 = ' + str(R_square)[0:5] + '%')
                            phe_dic['curve_temp_' + str(year)].setData(x_temp, seven_para_logistic_function(x_temp, paras[0], paras[1], paras[2], paras[3], paras[4], paras[5], paras[6]))
                            phe_dic['text_temp_' + str(year)].setText(msg_r_square)
                        else:
                            phe_dic['curve_temp_' + str(year)].setData(np.array([-2, -1]), np.array([-2, -1]))
                            phe_dic['text_temp_' + str(year)].setText('')
                        phase3_time += time.time() - phase3_s
                year_t += 1

**Comment by NilsNemitz** [comment]
Dear @shixi2ng ,

It's great to hear that things work for you now, and that pyqtgraph is of some help in handling the visualization of this amount of data in python!

If you run into any further problems, please feel free to open a new issue at any time.
We don't usually work with creating and exporting such a large number of separate curves, so this is a good stress test!
